### PR TITLE
HDDS-8829. Symmetric Keys for Delegation Tokens

### PR DESCRIPTION
**Pull request by ChenSammi**
## What changes were proposed in this pull request?
With secret key service provided by SCM, using the shared symmetric key to sign and verify delegation token. 

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8829

## How was this patch tested?

Existing delegation token tests, plus newly added compatibility unit tests.

### CODE REVIEW

**Comment by adoroszlai** [comment]
Please wait for clean CI run in fork before opening PR (or marking "ready for review").

https://github.com/ChenSammi/ozone/actions/runs/11678259968
https://github.com/ChenSammi/ozone/actions/runs/11678501108

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SecretKeyManagerService.java [hunk 1]
```diff
@@ -158,7 +158,6 @@ public void stop() {
   }
 
   public static boolean isSecretKeyEnable(SecurityConfig conf) {
-    return conf.isSecurityEnabled() &&
-        (conf.isBlockTokenEnabled() || conf.isContainerTokenEnabled());
+    return conf.isSecurityEnabled();
```
Comment (jojochuang) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SecretKeyManagerService.java:161]: why is this changed?

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java [hunk 2]
```diff
@@ -125,7 +136,13 @@ public OzoneTokenIdentifier fromUniqueSerializedKey(byte[] rawData)
...
     } else {
       this.tokenType = Type.DELEGATION_TOKEN;
-      setOmCertSerialId(WritableUtils.readString(in));
+      String value = WritableUtils.readString(in);
+      try {
+        UUID.fromString(value);
+        setSecretKeyId(value);
+      } catch (IllegalArgumentException e) {
+        setOmCertSerialId(value);
+      }
```
Comment (jojochuang) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java:145]: using exception handling for this case is probably going to be expensive.
what about:
```suggestion
      UUID.fromString(value);
      if (this.omCertSerialId == null || this.secretKeyId == null) {
        setSecretKeyId(value);
      } else {
        setOmCertSerialId(value);
      }
```

Comment (jojochuang) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java:145]: Oh wait this is not right. Why does it call UUID.fromString() and ignore the return value?

**Review by jojochuang — COMMENTED** [review]
will a new client work with an old OM?
In this case the client would fill both omCertSerialId and secretKeyId fields, OM would return a token based on the old implementation, so that keeps the old behavior, but at least that's not a regression. So it should work right?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/OzoneDelegationTokenSecretManager.java [hunk 3]
```diff
@@ -195,9 +209,15 @@ public Token<OzoneTokenIdentifier> createToken(Text owner, Text renewer,
     OzoneTokenIdentifier identifier = createIdentifier(owner, renewer,
         realUser);
     updateIdentifierDetails(identifier);
-
-    byte[] password = createPassword(identifier.getBytes(),
-        getCurrentKey().getPrivateKey());
+    byte[] password;
+    if (ozoneManager.getVersionManager().isAllowed(OMLayoutFeature.DELEGATION_TOKEN_SYMMETRIC_SIGN)) {
```
Comment (duongkame) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/OzoneDelegationTokenSecretManager.java:213]: We may not need a new layout version. This looks like a compatible change.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SecretKeyManagerService.java (hunk 1 above)
Comment (ChenSammi) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SecretKeyManagerService.java:161]: Before this patch, secretKey service only starts when block token or container token is enabled. With this patch, secretKey should start only when security is enabled, for delegation token is by default enabled when security is enabled.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java (hunk 2 above)
Comment (ChenSammi) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java:145]: Because this OzoneTokenIdentifier#toUniqueSerializedKey uses its own raw format instead of OMTokenProto format, which I have no idea how.  In a period of time(7d, max delegation time lifetime) after OM is upgraded to use shared secret key, the same index string would be cerial ID or secreKey ID. So this "  UUID.fromString(value);" used to detect whether it's secretKey, which is a UUID, or a serial ID, which is a number.

**Comment by ChenSammi** [comment]
> work



> will a new client work with an old OM? In this case the client would fill both omCertSerialId and secretKeyId fields, OM would return a token based on the old implementation, so that keeps the old behavior, but at least that's not a regression. So it should work right?

A new client work with an old OM will get an old RSA signed delegation token. And old RSA signed delegation token can be verified by both old OM and new OM.  It will work.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java [hunk 4]
```diff
@@ -326,6 +359,18 @@ public String getOmCertSerialId() {
 
   public void setOmCertSerialId(String omCertSerialId) {
     this.omCertSerialId = omCertSerialId;
+    Preconditions.checkArgument(this.omCertSerialId == null || this.secretKeyId == null,
```
[the remark this answers is not in the document]
Comment (ChenSammi) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java:362]: The check here is to make sure a OzoneTokenIdentifier cannot have both omCertSerialId  and secretKeyId set.  It can only have one of them.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/OzoneDelegationTokenSecretManager.java (hunk 3 above)
Comment (ChenSammi) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/OzoneDelegationTokenSecretManager.java:213]: The token signed by symmetric keys and token signed by asymmetric keys are incompatible at server side.  Delegation tokens are relatively long lived token, default 7d.  If a client get a delegation signed by symmetric keys after OM upgrade but before finalize, later this OM upgrade is cancelled, then when the token client held sent to the OM again after downgrade,  it cannot be verified by OM, client execution will fail.  All type of tokens are opaque to client.  The new DELEGATION_TOKEN_SYMMETRIC_SIGN layout is introduced to handle the compatibility between lower version of OM and high version of OM.

File: hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto [hunk 5]
```diff
@@ -1458,6 +1458,7 @@ message OMTokenProto {
     optional string signature      = 13;
     optional string strToSign      = 14;
     optional string omServiceId    = 15;
+    optional string secretKeyId    = 16;
```
[the remark this answers is not in the document]
Comment (ChenSammi) [reply hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto:1461]: Right

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java (hunk 2 above)
Comment (jojochuang) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/OzoneTokenIdentifier.java:115]: I was going to suggest to use Integer.isNumeric(). But it throws an exception internally so it ends up the same.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/OzoneDelegationTokenSecretManager.java [hunk 6]
```diff
@@ -436,6 +455,21 @@ private TokenInfo validateToken(OzoneTokenIdentifier identifier)
    */
   public boolean verifySignature(OzoneTokenIdentifier identifier,
       byte[] password) {
+    String secretKeyId = identifier.getSecretKeyId();
+    if (StringUtils.isNotEmpty(secretKeyId)) {
```
Comment (jojochuang) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/OzoneDelegationTokenSecretManager.java:459]: The PR overall looks good. The only thing I have an opinion about is six months from now, I won't remember why we have an if else statement here. How about making this an intuitive helper method, like 
```suggestion
    if (isSymmetricKey(identifier)) {
```

or simply add a comment here.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/OzoneDelegationTokenSecretManager.java [hunk 7]
```diff
@@ -528,8 +562,17 @@ private void addPersistedDelegationToken(OzoneTokenIdentifier identifier,
           "Can't add persisted delegation token to a running SecretManager.");
     }
 
-    byte[] password = createPassword(identifier.getBytes(),
-        getCertClient().getPrivateKey());
+    byte[] password;
+    if (StringUtils.isNotEmpty(identifier.getSecretKeyId())) {
```
Comment (jojochuang) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/OzoneDelegationTokenSecretManager.java:566]: Here, too.

**Review by jojochuang — APPROVED** [review]

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/upgrade/OMLayoutFeature.java [hunk 8]
```diff
@@ -46,7 +46,8 @@ public enum OMLayoutFeature implements LayoutFeature {
   FILESYSTEM_SNAPSHOT(5, "Ozone version supporting snapshot"),
 
   QUOTA(6, "Ozone quota re-calculate"),
-  HBASE_SUPPORT(7, "Full support of hsync, lease recovery and listOpenFiles APIs for HBase");
+  HBASE_SUPPORT(7, "Full support of hsync, lease recovery and listOpenFiles APIs for HBase"),
+  DELEGATION_TOKEN_SYMMETRIC_SIGN(8, "Delegation token signed by symmetric key");
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/upgrade/OMLayoutFeature.java:50]: I may be missing something, but why is this an `OMLayoutFeature`?  Is there any permanent change at the server-side that needs to be finalized after upgrade before we can start using symmetric keys?

**Comment by adoroszlai** [comment]
BTW, here is a simple update for the `xcompat` test to cover delegation tokens, would be nice to add it to the patch.

```diff
diff --git hadoop-ozone/dist/src/main/smoketest/compatibility/read.robot hadoop-ozone/dist/src/main/smoketest/compatibility/read.robot
index e56465bb8b..00f5408bc6 100644
--- hadoop-ozone/dist/src/main/smoketest/compatibility/read.robot
+++ hadoop-ozone/dist/src/main/smoketest/compatibility/read.robot
@@ -46,3 +46,8 @@ FSO Bucket Can Be Read
     Execute    ozone fs -get ofs://om/vol1/fso-bucket-${SUFFIX}/dir/subdir/file ${TEMP_DIR}/
     Execute    diff -q ${TESTFILE} ${TEMP_DIR}/file
     [teardown]    Execute    rm -f ${TEMP_DIR}/file
+
+Delegation Token Can Be Renewed
+    Execute    ozone sh key get -f /vol1/bucket1/token-${SUFFIX} ${TEMP_DIR}/token
+    Execute    ozone sh token renew --token ${TEMP_DIR}/token
+    [teardown]    Execute    rm -f ${TEMP_DIR}/token
diff --git hadoop-ozone/dist/src/main/smoketest/compatibility/write.robot hadoop-ozone/dist/src/main/smoketest/compatibility/write.robot
index a44901a99a..0453f00c0d 100644
--- hadoop-ozone/dist/src/main/smoketest/compatibility/write.robot
+++ hadoop-ozone/dist/src/main/smoketest/compatibility/write.robot
@@ -40,3 +40,8 @@ FSO Bucket Can Be Created and Used
     Execute    ozone sh bucket create --layout FILE_SYSTEM_OPTIMIZED /vol1/fso-bucket-${SUFFIX}
     Execute    ozone fs -mkdir -p ofs://om/vol1/fso-bucket-${SUFFIX}/dir/subdir
     Execute    ozone fs -put ${TESTFILE} ofs://om/vol1/fso-bucket-${SUFFIX}/dir/subdir/file
+
+Delegation Token Can Be Obtained
+    [setup]    Execute    rm -f ${TEMP_DIR}/token
+    Execute    ozone sh token get --token ${TEMP_DIR}/token
+    Execute    ozone sh key put /vol1/bucket1/token-${SUFFIX} ${TEMP_DIR}/token
```

I've run this with `old_versions="1.4.0"` only, assuming all prior releases have the same behavior for tokens.  Both `token get` and `token renew` works well.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/upgrade/OMLayoutFeature.java (hunk 8 above)
Comment (ChenSammi) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/upgrade/OMLayoutFeature.java:50]: The token signed by symmetric keys and token signed by asymmetric keys are incompatible at server side. Delegation tokens are relatively long lived token, default 7d. If a client get a delegation signed by symmetric keys after OM upgrade but before finalize, later this OM upgrade is cancelled, then when the token client held sent to the OM again after downgrade, it cannot be verified by OM, client execution will fail. All type of tokens are opaque to client. The new DELEGATION_TOKEN_SYMMETRIC_SIGN layout is introduced to handle the compatibility between lower version of OM and high version of OM.

Comment (adoroszlai) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/upgrade/OMLayoutFeature.java:50]: Thanks @ChenSammi for the great explanation, makes sense.

**Review by duongkame — APPROVED** [review]
LGTM. Thanks for the change @ChenSammi .

**Comment by adoroszlai** [comment]
Thanks @ChenSammi for the patch, @duongkame, @jojochuang for the review.

**Comment by ChenSammi** [comment]
Thanks @adoroszlai @jojochuang @duongkame for the review.   @adoroszlai , thanks for this xcompat test.
